### DISCUSSION TITLE
FSLeyes not opening in macOS Big Sur

### DISCUSSION THREAD

**Discussion by Felipe-RAndreis** [neuropoly/axondeepseg#570] [Show and tell]
Hello everyone,

I have installed AxonDeepSeg from the tutorial website (https://axondeepseg.readthedocs.io/en/latest/documentation.html#miniconda). I have not had any python related software previously installed on my computer.

**Describe the bug**
However, when I try to open fsleyes, the following error occurs:
/Users/username/miniconda/envs/ads_venv/bin/pythonw: line 3: 14101 Segmentation fault: 11  /Users/username/miniconda/envs/ads_venv/python.app/Contents/MacOS/python "$@"

Additionally, python crashes and an additional window opens with the following message: python quit unexpectedly.

**To reproduce**
$ conda activate ads_venv
$ fsleyes

**Platform**
macOS Big Sur version 11.6

**Comment by jcohenadad** [comment]
@Felipe-RAndreis maybe this would be helpful? https://spinalcordtoolbox.com/user_section/installation/fsleyes.html#macos-big-sur

**Reply by Felipe-RAndreis** [reply]
@jcohenadad Hi,

It helped partially. I still encounter problems. I have tried the proposed solution. The first option, which is "_When asked for which operating system you have, you will not see Big Sur...._".  This option did not appear to me. 

Therefore, I have tried the second option, which was to comment on a specific line and add 4 additional lines in the file named _ctypesloader.py_

When I load fsleyes and load the plugin (ads_plugin.py), the following error happens occurs: **Error installing plugin file** _An error occurred installing the plugin file. Details: No module named 'AxonDeepSeg'_

I could also notice how every time I open fsleyes, a new **cytpesloader.py** (which is the file to edit the lines) is created.

**Reply by jcohenadad** [reply]
@mathieuboudreau @alexfoias i think you have big sur, could you please help? thanks

**Reply by mathieuboudreau** [reply]
Looking into it

**Reply by mathieuboudreau** [reply]
Right – testing it on my laptop would complicate things and not really replicate @Felipe-RAndreis's situation, as I'm using a Mac M1 and to install ADS I need to run a separate hack, as Tensorflow isn't compatible.

@alexfoias can you test? Otherwise I'll give it a shot anyways.

**Reply by alexfoias** [reply]
yes, sure I will give it a try. As far as I can see the problem comes from FSLeyes and not from ADS.

**Reply by mathieuboudreau** [reply]
Just to be sure, @Felipe-RAndreis could you run `conda info fsleyes` from within your virtual environment, and paste the output here?

**Reply by alexfoias** [reply]
I have managed to run it. But maybe because I hacked the FSLeyes config everything seems to work well.

https://gist.github.com/alexfoias/0af02e5f125f2008ffefced5ab84fcaf

```
(ads_venv) [6/10/21 1:22:41] alex@Sansfil-Securise-Employes-Poly-229-250 axondeepseg % which fsleyes
/Users/alex/miniconda3/envs/ads_venv/bin/fsleyes
```

On the other hand I have this safety error when installing the venv:

```
SafetyError: The package for pyopengl located at /Users/alex/miniconda3/pkgs/pyopengl-3.1.5-py_0
appears to be corrupted. The path 'site-packages/OpenGL/platform/ctypesloader.py'
has an incorrect size.
  reported size: 3402 bytes
  actual size: 3475 bytes
```

**Reply by mathieuboudreau** [reply]
> I have managed to run it. But maybe because I hacked the FSLeyes config everything seems to work well.
> 
> https://gist.github.com/alexfoias/0af02e5f125f2008ffefced5ab84fcaf
> 
> ```
> (ads_venv) [6/10/21 1:22:41] alex@Sansfil-Securise-Employes-Poly-229-250 axondeepseg % which fsleyes
> /Users/alex/miniconda3/envs/ads_venv/bin/fsleyes
> ```
> 
> On the other hand I have this safety error when installing the venv:
> 
> ```
> SafetyError: The package for pyopengl located at /Users/alex/miniconda3/pkgs/pyopengl-3.1.5-py_0
> appears to be corrupted. The path 'site-packages/OpenGL/platform/ctypesloader.py'
> has an incorrect size.
>   reported size: 3402 bytes
>   actual size: 3475 bytes
> ```

Ha, I'd backup my computer just in case it's a failing hard drive =)

**Reply by alexfoias** [reply]
My `ctypesloader.py` looks like this:

```
(ads_venv) [6/10/21 1:25:57] alex@Sansfil-Securise-Employes-Poly-229-250 pyopengl-3.1.5-py_0 % more site-packages/OpenGL/platform/ctypesloader.py 
"""ctypes abstraction layer

We keep rewriting functions as the main entry points change,
so let's just localise the changes here...
"""
import ctypes, logging, os, sys
_log = logging.getLogger( 'OpenGL.platform.ctypesloader' )
#_log.setLevel( logging.DEBUG )
ctypes_version = [
    int(x) for x in ctypes.__version__.split('.')
]
from ctypes import util
import OpenGL

DLL_DIRECTORY = os.path.join( os.path.dirname( OpenGL.__file__ ), 'DLLS' )

def loadLibrary( dllType, name, mode = ctypes.RTLD_GLOBAL ):
    """Load a given library by name with the given mode
    
    dllType -- the standard ctypes pointer to a dll type, such as
        ctypes.cdll or ctypes.windll or the underlying ctypes.CDLL or 
        ctypes.WinDLL classes.
    name -- a short module name, e.g. 'GL' or 'GLU'
    mode -- ctypes.RTLD_GLOBAL or ctypes.RTLD_LOCAL,
        controls whether the module resolves names via other
        modules already loaded into this process.  GL modules
        generally need to be loaded with GLOBAL flags
    
    returns the ctypes C-module object
    """
    if isinstance( dllType, ctypes.LibraryLoader ):
        dllType = dllType._dlltype
    if sys.platform.startswith('linux'):
        return _loadLibraryPosix(dllType, name, mode)
    else:
        return _loadLibraryWindows(dllType, name, mode)


def _loadLibraryPosix(dllType, name, mode):
    """Load a given library for posix systems

    The problem with util.find_library is that it does not respect linker runtime variables like
    LD_LIBRARY_PATH.

    Also we cannot rely on libGLU.so to be available, for example. Most of Linux distributions will
    ship only libGLU.so.1 by default. Files ending with .so are normally used when compiling and are
    provided by dev packages.

    returns the ctypes C-module object
    """
    prefix = 'lib'
    suffix = '.so'
    base_name = prefix + name + suffix
    
    filenames_to_try = []
    # If a .so is missing, let's try libs with so version (e.g libGLU.so.9, libGLU.so.8 and so on)
    filenames_to_try.extend(list(reversed([
        base_name + '.%i' % i for i in range(0, 10)
    ])))
    err = None

    for filename in filenames_to_try:
        try:
            result = dllType(filename, mode)
            _log.debug( 'Loaded %s => %s %s', base_name, filename, result)
            return result
        except Exception as current_err:
            err = current_err
    
    _log.info('''Failed to load library ( %r ): %s''', filename, err or 'No filenames available to guess?')

def _loadLibraryWindows(dllType, name, mode):
    """Load a given library for Windows systems

    returns the ctypes C-module object
    """
    fullName = None
    try:
#        fullName = util.find_library( name )
        fullName = '/System/Library/Frameworks/OpenGL.framework/OpenGL'
        if fullName is not None:
            name = fullName
        elif os.path.isfile( os.path.join( DLL_DIRECTORY, name + '.dll' )):
            name = os.path.join( DLL_DIRECTORY, name + '.dll' )
    except Exception as err:
        _log.info( '''Failed on util.find_library( %r ): %s''', name, err )
        # Should the call fail, we just try to load the base filename...
        pass
    try:
        return dllType( name, mode )
    except Exception as err:
        err.args += (name,fullName)
        raise

def buildFunction( functionType, name, dll ):
    """Abstract away the ctypes function-creation operation"""
    return functionType( (name, dll), )

```

**Reply by Felipe-RAndreis** [reply]
Just a quick question before pasting the output: Do I need to install fsleyes inside the AxonDeepSeg folder?

> Just to be sure, @Felipe-RAndreis could you run `conda info fsleyes` from within your virtual environment, and paste the output here?

Hey, Here's the output:
(ads_venv) user$ conda info fsleyes
WARNING: 'conda info package_name' is deprecated.
          Use 'conda search package_name --info'.
(ads_venv) user$ conda search fsleyes --info
Loading channels: done
No match found for: fsleyes. Search: *fsleyes*

PackagesNotFoundError: The following packages are not available from current channels:

  - fsleyes

Current channels:

  - https://repo.anaconda.com/pkgs/main/osx-64
  - https://repo.anaconda.com/pkgs/main/noarch
  - https://repo.anaconda.com/pkgs/r/osx-64
  - https://repo.anaconda.com/pkgs/r/noarch

To search for alternate channels that may provide the conda package you're
looking for, navigate to

    https://anaconda.org

and use the search bar at the top of the page.

**Reply by Felipe-RAndreis** [reply]
> > I have managed to run it. But maybe because I hacked the FSLeyes config everything seems to work well.
> > https://gist.github.com/alexfoias/0af02e5f125f2008ffefced5ab84fcaf
> > ```
> > (ads_venv) [6/10/21 1:22:41] alex@Sansfil-Securise-Employes-Poly-229-250 axondeepseg % which fsleyes
> > /Users/alex/miniconda3/envs/ads_venv/bin/fsleyes
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > On the other hand I have this safety error when installing the venv:
> > ```
> > SafetyError: The package for pyopengl located at /Users/alex/miniconda3/pkgs/pyopengl-3.1.5-py_0
> > appears to be corrupted. The path 'site-packages/OpenGL/platform/ctypesloader.py'
> > has an incorrect size.
> >   reported size: 3402 bytes
> >   actual size: 3475 bytes
> > ```
> 
> Ha, I'd backup my computer just in case it's a failing hard drive =)

Hi, when I try to use which fsleyes this is returned.

(ads_venv) user:axondeepseg user$ which fsleyes
/usr/local/fsl/bin/fsleyes

**Reply by mathieuboudreau** [reply]
Ah, it appears that the FSL installation that your path sees is not in your AxonDeepSeg virtual environment. Could you explain how you installed ADS - do you have the latest (master) version? Did you have FSL installed prior or did you try and install it seperately? It's possible that it's added in your shell path file after your venv, if you installed FSL through their instruction after installing ADS with conda.

**Reply by Felipe-RAndreis** [reply]
> Ah, it appears that the FSL installation that your path sees is not in your AxonDeepSeg virtual environment. Could you explain how you install ADS - do you have the latest (master) version? Did you have FSL installed prior or did you try and install it seperately? It's possible that it's added in your shell path file after your venv, if you installed FSL through their instruction after installing ADS with conda.

Hi Mathieu,

**Could you explain how you install ADS:** I followed these instructions: https://axondeepseg.readthedocs.io/en/latest/documentation.html#installation

**do you have the latest (master) version?** I don't know if I have the latest (master) version. How can I check this? I downloaded yesterday from this website: https://axondeepseg.readthedocs.io/en/latest/documentation.html#installation

**Did you have FSL installed prior or did you try and install it seperately?**: I did not have FSL installed prior. I installed after installing ADS. I have installed through here: https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/FslInstallation/MacOsX. 
Basically, I have downloaded **fslinstaller.py** and then just typed:
cd ~/Downloads
python fslinstaller.py

I have also tried through here: **Install from conda-forge (recommended)**
FSLeyes is available on conda-forge - if you use an anaconda or miniconda environment, you can install FSLeyes into it like so:
conda install -c conda-forge fsleyes
Or you can create a separate conda environment, specifically for FSLeyes, like so:
conda create -c conda-forge -p path/to/fsleyes/environment fsleyes

I have downloaded both ways.

**It's possible that it's added in your shell path file after your venv, if you installed FSL through their instruction after installing ADS with conda.**: I don't understand how should I do it correctly.

**What do you think about completely removing FSLeyes and installing through the 'correct' way?** I just don't know which is the correct one. Should I install fsleyes while in the ads_venv or in the base?

Finally, I can see Innside user/miniconda/envs/ads_venv/bin there are fsl files.

**PS:** I just remembered that when installing FSL, it is asked: _Where would you like the FSL install to be (including the FSL folder name)? [/usr/local/fsl]:_ I have used this option.

**Reply by mathieuboudreau** [reply]
Ah ha, yes I think I understand (one of) the problem(s) now. 

You didn't need to install FSL; I think doing so, it added it in your shell path file in such a way that it supersedes your virtual environment. FSLeyes gets install during the AxonDeepSeg installation, and it does so in such a way that FSLeyes "sees" all the ADS requirements/environment/ADS installation.

If you could try running `/user/miniconda/envs/ads_venv/bin/fsleyes` from the command line (while inside your venv), it might open the correct fsleyes version, and you may or may not have a functionning version of FSLeyes there (you'll still need to add the plugin though).

**Reply by Felipe-RAndreis** [reply]
> Ah ha, yes I think I understand (one of) the problem(s) now.
> 
> You didn't need to install FSL; I think doing so, it added it in your shell path file in such a way that it supersedes your virtual environment. FSLeyes gets install during the AxonDeepSeg installation, and it does so in such a way that FSLeyes "sees" all the ADS requirements/environment/ADS installation.
> 
> If you could try running `/user/miniconda/envs/ads_venv/bin/fsleyes` from the command line (while inside your venv), it might open the correct fsleyes version, and you may or may not have a functionning version of FSLeyes there (you'll still need to add the plugin though).

Hi again,

So, I deleted all FSL folders outside the envs/ads_venv. Now, when I type which fsleyes (with venv activated), the result is:
**/Users/user/miniconda/envs/ads_venv/bin/fsleyes**

Then, when running _fsleyes_, it returns an error:

Traceback (most recent call last):
  File "/Users/fran/miniconda/envs/ads_venv/bin/fsleyes", line 11, in <module>
    sys.exit(main())
  File "/Users/fran/miniconda/envs/ads_venv/lib/python3.7/site-packages/fsleyes/filtermain.py", line 205, in main
    result = fm.main(args)
  File "/Users/fran/miniconda/envs/ads_venv/lib/python3.7/site-packages/fsleyes/main.py", line 243, in main
    fsleyes.initialise()
  File "/Users/fran/miniconda/envs/ads_venv/lib/python3.7/site-packages/fsleyes/__init__.py", line 243, in initialise
    import fsleyes.plugins as plugins
  File "/Users/fran/miniconda/envs/ads_venv/lib/python3.7/site-packages/fsleyes/plugins/__init__.py", line 196, in <module>
    import fsleyes.views.canvaspanel     as canvaspanel
  File "/Users/fran/miniconda/envs/ads_venv/lib/python3.7/site-packages/fsleyes/views/canvaspanel.py", line 20, in <module>
    from . import                    colourbarpanel
  File "/Users/fran/miniconda/envs/ads_venv/lib/python3.7/site-packages/fsleyes/views/colourbarpanel.py", line 15, in <module>
    import fsleyes.gl.wxglcolourbarcanvas as cbarcanvas
  File "/Users/fran/miniconda/envs/ads_venv/lib/python3.7/site-packages/fsleyes/gl/wxglcolourbarcanvas.py", line 17, in <module>
    import fsleyes.gl.colourbarcanvas as cbarcanvas
  File "/Users/fran/miniconda/envs/ads_venv/lib/python3.7/site-packages/fsleyes/gl/colourbarcanvas.py", line 21, in <module>
    import OpenGL.GL as gl
  File "/Users/fran/miniconda/envs/ads_venv/lib/python3.7/site-packages/OpenGL/GL/__init__.py", line 3, in <module>
    from OpenGL import error as _error
  File "/Users/fran/miniconda/envs/ads_venv/lib/python3.7/site-packages/OpenGL/error.py", line 12, in <module>
    from OpenGL import platform, _configflags
  File "/Users/fran/miniconda/envs/ads_venv/lib/python3.7/site-packages/OpenGL/platform/__init__.py", line 36, in <module>
    _load()
  File "/Users/fran/miniconda/envs/ads_venv/lib/python3.7/site-packages/OpenGL/platform/__init__.py", line 27, in _load
    plugin_class = plugin.load()
  File "/Users/fran/miniconda/envs/ads_venv/lib/python3.7/site-packages/OpenGL/plugins.py", line 16, in load
    return importByName( self.import_path )
  File "/Users/fran/miniconda/envs/ads_venv/lib/python3.7/site-packages/OpenGL/plugins.py", line 38, in importByName
    module = __import__( ".".join(moduleName), {}, {}, moduleName)
  File "/Users/fran/miniconda/envs/ads_venv/lib/python3.7/site-packages/OpenGL/platform/darwin.py", line 25, in <module>
    from OpenGL.platform import baseplatform, ctypesloader
  File "/Users/fran/miniconda/envs/ads_venv/lib/python3.7/site-packages/OpenGL/platform/baseplatform.py", line 4, in <module>
    from OpenGL.platform import ctypesloader
  File "/Users/fran/miniconda/envs/ads_venv/lib/python3.7/site-packages/OpenGL/platform/ctypesloader.py", line 80
    fullName = '/System/Library/Frameworks/OpenGL.framework/OpenGL'
                                                                  ^
TabError: inconsistent use of tabs and spaces in indentation

**Reply by mathieuboudreau** [reply]
Earlier in the thread, did you follow the directions here, resulting in editing the file ctypesloader.py? https://spinalcordtoolbox.com/user_section/installation/fsleyes.html#macos-big-sur?

It appears to be telling you that the tab on that line (or the line before) is not consistent with Python's required syntax. Make sure there are four "space" charaters preceeding `fullName`. If it's fine, then it means maybe the line before it has an extra space/space missing?

**Reply by Felipe-RAndreis** [reply]
> Earlier in the thread, did you follow the directions here, resulting in editing the file ctypesloader.py? https://spinalcordtoolbox.com/user_section/installation/fsleyes.html#macos-big-sur?
> 
> It appears to be telling you that the tab on that line (or the line before) is not consistent with Python's required syntax. Make sure there are four "space" charaters preceeding `fullName`. If it's fine, then it means maybe the line before it has an extra space/space missing?

Hi,

So, I copied @alexfoias _ctypesloader.py_ and now a different error occurs, which is: _/Users/fran/miniconda/envs/ads_venv/bin/pythonw: line 3: 86899 Segmentation fault: 11  /Users/fran/miniconda/envs/ads_venv/python.app/Contents/MacOS/python "$@"_

I have noticed that here (https://spinalcordtoolbox.com/user_section/installation/fsleyes.html#macos-big-sur), the instructions are to add 4 lines:
_if name == "OpenGL":
  fullName = "/System/Library/Frameworks/OpenGL.framework/OpenGL"
elif name == "GLUT":
  fullName = "/System/Library/Frameworks/GLUT.framework/GLUT"_

However, in his file, there's only the second line.
When I try to add these four lines, the if and elif I leave with four spaces. The other two lines, starting with _fullName_ I inserted with 6 spaces.

Then, an error occurs: if name == "OpenGL":
     ^
IndentationError: expected an indented block

This part of the file looks like:
#       fullName = util.find_library( name ) (this is commented with #)
    if name == "OpenGL":
      fullName = "/System/Library/Frameworks/OpenGL.framework/OpenGL"
    elif name == "GLUT":
      fullName = "/System/Library/Frameworks/GLUT.framework/GLUT"
        if fullName is not None:
            name = fullName
        elif os.path.isfile( os.path.join( DLL_DIRECTORY, name + '.dll' )):
            name = os.path.join( DLL_DIRECTORY, name + '.dll' )
